src/main/java/ch/jalu/configme/migration/VersionMigrationService.java

```diff
@@ -0,0 +1,236 @@
+        if (currentConfigVersion > getCurrentVersion()) handleTooHigherVersion(currentConfigVersion);
+
+        return performMigrations(reader, configurationData) || !configurationData.areAllValuesValidInResource();
+    }
+
+    /**
+     * @return the start config version. By default, it is 1.
+     */
+    public int getStartVersion() {
+        return this.startVersion;
+    }
+
+    /**
+     * @return the current config version.
+     */
+    public int getCurrentVersion() {
+        return this.currentVersion;
+    }
+
+    /**
+     * @return The config property name. By default, it is "version".
+     */
+    @NotNull
+    public String getVersionPropertyName() {
+        return this.versionPropertyName;
+    }
+
+    /**
+     * @return The unmodifiable {@link Collection<Migration>} of migrations.
+     */
+    @NotNull
+    public Collection<Migration> getMigrations() {
+        return this.migrations;
+    }
+
+    /**
+     * Invoked if the version is lower than {@link #getStartVersion()}.
+     */
+    protected void handleTooLowerVersion(int version) {
+        throw new IllegalStateException("The current version (" + version + ") is lower than " + getStartVersion());
+    }
+
+    /**
+     * Invoked if the version is greater than {@link #getCurrentVersion()}.
+     */
+    protected void handleTooHigherVersion(int version) {
+        throw new IllegalStateException("The current version (" + version + ") is greater than " + getCurrentVersion());
+    }
+
+    /**
+     * Perform the migration by using the versioning system.
+     * <p>
+     * Note that the settings manager automatically saves the resource
+     * if the migration service returns {@link #MIGRATION_REQUIRED} from {@link #checkAndMigrate}.
+     *
+     * @param reader the reader with which the configuration file can be read
+     * @param configurationData the configuration data
+     * @return true if a migration has been performed, false otherwise (see constants on {@link MigrationService})
+     */
+    protected boolean performMigrations(@NotNull PropertyReader reader, @NotNull ConfigurationData configurationData) {
+        int currentConfigVersion = readConfigurationVersion(reader, configurationData);
+        AtomicBoolean migrationResult = new AtomicBoolean(NO_MIGRATION_NEEDED);
+
+        // Migrate the configuration from the version 1 to 2 to 3 to ...
+        migrations
+            .forEach(migration -> {
+                if (migration == null) return; // null-safe check.
+
+                int fromVersion = migration.fromVersion();
+                int toVersion = migration.toVersion();
+
+                // start the migration.
+                if (currentConfigVersion == fromVersion) {
+                    migration.migrate(reader, configurationData);
+                    setConfigurationVersion(configurationData, toVersion);
+                    migrationResult.set(MIGRATION_REQUIRED);
+                }
+            });
```

Comment:
I'm wondering how we can make this the most robust. If I have multiple migrations from one version to another, only one will be run. The order of the migration is somehow relevant but the migrations are held in a `Collection`, which could be a HashSet or anything else with non-deterministic values.

The question is a bit what we want to offer to the user?
* Should it be possible to pass in multiple migrations from version 1 to 2? If I have a lot of things I'm changing, I might want to put them into separate migrations. 
* If I have a migration 1->3 and 2->3, do I expect the migration 2->3 to be run? The code currently says no, and I would agree with that: the migration explicitly says it migrates from 1 to 3, so a migration _from_ 2 is never relevant.
* What if my current config version is 4, and I only have migrations from 1 to 2, and from 3 to 4? I don't think it would be correct to just assume that after 2 we can run the migration 3->4, and indeed your current code doesn't do that. I think this is good. After all, there's probably not much point to increase the version if there's no migration, but we'll have to keep that in mind and see if we can communicate that to the user a bit more :smile:
* Maybe validating that a migration's `toVersion` is not larger than the application's current version would be wise

I think the easiest to make this "deterministic" is to allow only one migration per `fromVersion` and to keep that in a `Map<Integer, Migration>` (the key being the `fromVersion`). If there is a migration for that version, run it, if there is a migration for the new version, run it, and so forth.